File: scala-native/src/main/scala/qt/core/QByteArray.scala
package qt.core

import scalanative.unsafe.*
import scala.util.Using.Releasable
import scala.util.Using.Manager

opaque type QByteArray = Ptr[Byte]

@extern
@link("Qt5Core")
private object _QByteArray:
  def q_byte_array_init(data: CString): QByteArray = extern
  def q_byte_array_destroy(array: QByteArray): Unit = extern

object QByteArray:
  given Releasable[Zone] with 
    def release(resource: Zone): Unit = resource.close()
  given Releasable[QByteArray] with
    def release(resource: QByteArray): Unit =
      println("destroying byte array")
      _QByteArray.q_byte_array_destroy(resource)

  def apply(data: String)(using m: Manager) = 
    println("creating qbytearray")
    given Zone = m(Zone.open())
    m(_QByteArray.q_byte_array_init(toCString(data)))

File: scala-native/src/main/resources/scala-native/qt/core/QByteArray.cpp
#include <QByteArray>
#include "QByteArray.hpp"

q_byte_array q_byte_array_init(const char* data) {
  return new QByteArray(data);
}

void q_byte_array_destroy(q_byte_array array) {
  delete (QByteArray*) array;
  return;
}